aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
-  Tensor ret = at::empty(self.sizes(), self.options());
+  // Here we do the type promotion without the TensorIterator machinery.
+  // For `where` we only want to consider `self` and `other`,
+  // however right now TensorIterator API doesn't have a way to specify which inputs
+  // should take part in type promotion.
+  auto common_dtype = at::result_type(self, other);
```

Comment:
`at::result_type` is also implemented for (scalar, tensor) and (tensor, scalar) args. Would it make more sense to make use of that rather than using `wrapped_scalar_tensor_default_dtype` (which is a bit convoluted and also seems to have some comment saying not to use it)?

Reply:
In that case we will need to add new overload for `where` kernel which takes a scalar as parameter (because the current version expects all inputs to be passed via TensorIterator which only takes Tensor inputs)  and add other overload for `_s_where` and corresponding entry in `derivatives.yaml`

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L204-L205